context.go

```diff
@@ -0,0 +1,44 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package httprequest
+
+import (
+	"net/http"
+
+	"github.com/juju/utils"
+	"golang.org/x/net/context"
+)
+
+const requestUUIDHeader = "X-Request-UUID"
+
+type requestUUIDContextKey struct{}
+
+// RequestUUID returns the unique identifier of the request. This will
+// have either been taken from a X-Request-UUID header or assigned when
+// the request is initially processed by httprequest. If the given
+// context doesn't contain a request UUID then the return value will be
+// the empty string.
+func RequestUUID(ctx context.Context) string {
+	v := ctx.Value(requestUUIDContextKey{})
+	if v == nil {
+		return ""
+	}
+	return v.(string)
+}
+
+// setRequestUUID adds a request UUID to the given context. The request
+// UUID either comes from a X-Request-UUID header in the given Request,
+// or generates a random UUID.
+func setRequestUUID(ctx context.Context, req *http.Request) context.Context {
+	uuid := req.Header.Get(requestUUIDHeader)
+	if uuid == "" {
+		u, err := utils.NewUUID()
```

Comment:
I'd suggest that instead of using utils.NewUUID which uses real random data,
we could use github.com/rogpeppe/fastuuid which will impose less overhead
and can't fail after initialization.

If we really want the conventional formatting for UUIDs (I personally prefer
just a string of hex or base64), then this would work:

    var uuid utils.UUID
    uuid0 := uuidGen.Next()
    copy(uuid[:], uuid0[:])
    uuid.String()